rio_tiler/tasks.py

```diff
@@ -1,86 +1,138 @@
-    Attributes
-    ----------
-    tasks: list or generator
-        Sequence of 'concurrent.futures._base.Future' or 'callable'
-    allowed_exceptions: Tuple, optional
-        List of exceptions which won't be raised.
-
-    Yields
-    ------
-    Successful task's result
-
-    """
-    if not allowed_exceptions:
-        allowed_exceptions = ()
-
-    for future, asset in tasks:
-        try:
-            if isinstance(future, futures.Future):
-                yield future.result(), asset
-            else:
-                yield future, asset
-        except allowed_exceptions as err:
-            logging.info(err)
-            pass
 
 
-def create_tasks(reader: Callable, assets, threads, *args, **kwargs) -> TaskType:
-    """Create Future Tasks."""
-    if threads and threads > 1:
-        with futures.ThreadPoolExecutor(max_workers=threads) as executor:
+@attr.s
+class TaskManager(abc.ABC):
+    """Task manager base class"""
+
+    max_threads: int = attr.ib()
+
+    @classmethod
+    def create_from_env(cls):
+        """Create from environment."""
+        if MAX_THREADS == 1:
+            return SingleThreadedManager()
+        else:
+            return MultiThreadedManager(max_threads=MAX_THREADS)
+
+    @abc.abstractmethod
+    def create_tasks(self, reader: Callable, assets, *args, **kwargs) -> Iterable:
+        """Create Future Tasks."""
+        ...
+
+    @abc.abstractmethod
+    def filter_tasks(self, tasks: Iterable, allowed_exceptions: Tuple = ()) -> Iterable:
+        """Filter tasks to remove Exceptions."""
+
+
+@attr.s
+class SingleThreadedManager(TaskManager):
+    """Single threaded task management"""
+
+    max_threads: int = 1
+
+    def create_tasks(
+        self, reader: Callable, assets, *args, **kwargs
+    ) -> Generator[Tuple[Callable, str], None, None]:
+        """Create Future Tasks."""
+        return ((reader(asset, *args, **kwargs), asset) for asset in assets)
+
+    def filter_tasks(
+        self, tasks: Iterable, allowed_exceptions: Tuple = ()
+    ) -> Generator:
+        """
+        Filter tasks to remove Exceptions.
+
+        Attributes
+        ----------
+        tasks: iterable of 'callable'
+            Iterable of 'concurrent.futures._base.Future' or 'callable'
+        allowed_exceptions: Tuple, optional
+            List of exceptions which won't be raised.
+
+        Yields
+        ------
+        Successful task's result
+
+        """
+        for ret, asset in tasks:
+            try:
+                yield ret, asset
+            except allowed_exceptions as err:
+                logging.info(err)
+                pass
+
+
+@attr.s
+class MultiThreadedManager(TaskManager):
+    """Multi threaded task management"""
+
+    max_threads: int
+
+    def create_tasks(
+        self, reader: Callable, assets, *args, **kwargs
+    ) -> Sequence[Tuple[futures.Future, str]]:
+        """Create Future Tasks."""
+        with futures.ThreadPoolExecutor(max_workers=self.max_threads) as executor:
```

Comment:
executor could be defined as a ClassVar if we wanted to re-use the same executor